### PR TITLE
Add libftdi versions 1.2-1.5.

### PR DESCRIPTION
**Pull request by planetmarshall**
Specify library name and version:  **lib/1.0**

- [x ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libftdi/1.x/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        os.unlink(os.path.join(self.package_folder, "bin", "libftdi1-config"))
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "LibFTDI1"
+        self.cpp_info.names["cmake_find_package_multi"] = "LibFTDI1"
+        self.cpp_info.names["pkgconfig"] = "libftdi1"
+        self.cpp_info.includedirs.append(
+            os.path.join(self.package_folder, "include", "libftdi1")
+        )
+        self.cpp_info.libs = ["ftdipp1", "ftdi1"]
```
Comment (madebr) [inline recipes/libftdi/1.x/conanfile.py:73]: You combine both libraries.
Does the cmake config file, installed by libftdi's cmake config (and removed by us), create separate targets for the c and c++ library?
Maybe this requires conan components?

File: recipes/libftdi/1.x/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        os.unlink(os.path.join(self.package_folder, "bin", "libftdi1-config"))
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "LibFTDI1"
+        self.cpp_info.names["cmake_find_package_multi"] = "LibFTDI1"
+        self.cpp_info.names["pkgconfig"] = "libftdi1"
+        self.cpp_info.includedirs.append(
+            os.path.join(self.package_folder, "include", "libftdi1")
+        )
```
Comment (madebr) [inline recipes/libftdi/1.x/conanfile.py:72]: No need for absolute paths.
`.includedirs` can contain relative paths.
`self.deps_cpp_info[xxx].include_paths` automatically converts these to absolute paths.
```suggestion
        self.cpp_info.includedirs.append(os.path.join("include", "libftdi1"))
```

File: recipes/libftdi/1.x/conandata.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    sha256: ec36fb49080f834690c24008328a5ef42d3cf584ef4060f3a35aa4681cb31b74
+    url: "https://www.intra2net.com/en/developer/libftdi/download/libftdi1-1.4.tar.bz2"
+
+  "1.3":
+    url: "https://www.intra2net.com/en/developer/libftdi/download/libftdi1-1.3.tar.bz2"
+    sha256: 9a8c95c94bfbcf36584a0a58a6e2003d9b133213d9202b76aec76302ffaa81f4
+
+  "1.2":
+    url: "https://www.intra2net.com/en/developer/libftdi/download/libftdi1-1.2.tar.bz2"
+    sha256: a6ea795c829219015eb372b03008351cee3fb39f684bff3bf8a4620b558488d6
```
Comment (madebr) [inline recipes/libftdi/1.x/conandata.yml:16]: Do we need all versions?
Isn't 1.5 enough?

I propose only adding 1.5.
If somebody really needs an older version, he/she can open a pr.

File: recipes/libftdi/1.x/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/libftdi/1.x/conanfile.py:71]: These batch of errors (the .pc related errors) show that you need to use components, 
and set the `pkg_config` names.

Comment (planetmarshall) [reply recipes/libftdi/1.x/conanfile.py]: I'm not sure it's possible to exactly mimic LibFTDI's CMake configuration with conan (Libftdi uses `LibFTDI1` as the package name but `LIBFTDI` and `LIBFTDIPP` as the package variable names), but I've attempted to get as close as I can with conan components using the targets `LibFTDI1::ftdi` and `LibFTDI1::ftdipp`.

Comment (prince-chrismc) [reply recipes/libftdi/1.x/conanfile.py]: Conan has 3 settings

Filenames (sets the cmake package name, is find_package)
Names (the cmake namespace of the targets)
Components names (each unique target)

I am sure we have an example I just can't recall one.

Comment (planetmarshall) [reply recipes/libftdi/1.x/conanfile.py]: The `find_package` name is `LibFTDI1`, as used in `examples/cmake_example`. The built-in find package then generates the variables `$LIBFTDI_LIBRARIES` and `$LIBFTDIPP_LIBRARIES`,

However the variables generated by the conan-generated find package when using components are `${LibFTDI1_component_LIBRARIES}` etc. I'm not sure it's possible to create multiple top-level variable names with conan, with a different name from the `find_package` (I did try various combinations as you suggested).

The targets created by LibFTDI are not namespaced, which I know conan doesn't support. See also conan-io/conan#7615

Comment (prince-chrismc) [reply recipes/libftdi/1.x/conanfile.py]: Yes since the variables fashion is outdated its unlikely that conan will add support for that. So yes it seems this can not be done.

File: recipes/libftdi/1.x/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            raise ConanInvalidConfiguration("Building with Visual Studio is not supported")
+
+    def build_requirements(self):
+        if not tools.which("pkg-config"):
+            self.build_requires("pkgconf/1.7.3")
+
+    def _patch_cmakelists(self, subfolder):
+        cmakelists_path = os.path.join(self._source_subfolder, subfolder, "CMakeLists.txt")
+        tools.replace_in_file(cmakelists_path, "CMAKE_SOURCE_DIR", "PROJECT_SOURCE_DIR", strict=False)
+        tools.replace_in_file(cmakelists_path, "CMAKE_BINARY_DIR", "PROJECT_BINARY_DIR", strict=False)
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/conanfile.py:50]: Looking over the project. I think we can do sometime much less aggressive. 

Can you please try to add the following? This way we can contribute it back to upstream 🤞 

[cmake-allow-subproject.patch.txt](https://github.com/conan-io/conan-center-index/files/5743632/cmake-allow-subproject.patch.txt)

File: recipes/libftdi/1.x/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    def package(self):
+        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+        lib_folder = os.path.join(self.package_folder, "lib",)
+        tools.rmdir(os.path.join(lib_folder, "cmake"))
+        tools.rmdir(os.path.join(lib_folder, "pkgconfig"))
+        # STATIC config creates both static and shared libraries
+        if not self.options.shared:
+            tools.remove_files_by_mask(lib_folder, "*.so*")
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/conanfile.py:84]: https://docs.conan.io/en/latest/changelog.html#id127

Please add [required_conan_version](https://github.com/conan-io/docs/pull/1788)

**Review by prince-chrismc — CHANGES_REQUESTED** [review]
Please fix the license 

I would allow appreciate if you could try the patch but it's not required to merge

**Comment by madebr** [comment]
@planetmarshall 
Please consider https://github.com/planetmarshall/conan-center-index/pull/1

**Comment by SSE4** [comment]
@planetmarshall ping :)

**Comment by planetmarshall** [comment]
> @planetmarshall ping :)

Thanks - I haven't forgotten :) I've forked the libftdi repository and created a branch for modernizing cmake usage, and a branch for adding MSVC support. Once complete I'll add patches from these branches to the conanfile. See https://github.com/planetmarshall/libftdi for details.

File: recipes/libftdi/1.x/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            "EXAMPLES": False,
+            "FTDI_EEPROM": False,
+            "FTDIPP" : True,
+            "STATICLIBS": not self.options.shared
+        }
+        self._cmake.definitions.update(options)
+        self._cmake.configure()
+        return self._cmake
+
+    def requirements(self):
```
Comment (madebr) [inline recipes/libftdi/1.x/conanfile.py:65]: Does this need libconfuse?

File: recipes/libftdi/1.x/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        self.cpp_info.names["pkgconfig"] = "libftdi1"
+
+        self.cpp_info.components["ftdi"].names["pkg_config"] = "libftdi1"
+        self.cpp_info.components["ftdi"].libs = ["ftdi1"]
+        self.cpp_info.components["ftdi"].requires = ["libusb::libusb"]
+        self.cpp_info.components["ftdi"].includedirs.append(os.path.join("include", "libftdi1"))
+
+        self.cpp_info.components["ftdipp"].names["pkg_config"] = "libftdi1pp"
+        self.cpp_info.components["ftdipp"].libs = ["ftdipp1"]
+        self.cpp_info.components["ftdipp"].requires = ["ftdi", "boost::boost"]
```
Comment (madebr) [inline recipes/libftdi/1.x/conanfile.py:96]: I'm not sure about this. What component of boost does libftdi use?
If it's e.g. filesystem, then you can only add `boost::filesystem`.
```suggestion
        self.cpp_info.components["ftdipp"].requires = ["ftdi", "boost::headers"]
```

File: recipes/libftdi/1.x/conanfile.py (hunk 6 above)
Comment (planetmarshall) [reply recipes/libftdi/1.x/conanfile.py:65]: libconfuse is only required by the eeprom utility, not by the library.

**Comment by prince-chrismc** [comment]
Missing MacOS framework symbols? https://github.com/ocornut/imgui/pull/2546#issuecomment-506042365

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 8]
```diff
@@ -0,0 +1,787 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 5aecafc..aa69995 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -1,36 +1,54 @@
+-# Project
+ project(libftdi1 C)
++
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:8]: extra

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 9]
```diff
@@ -0,0 +1,787 @@
...
++list(PREPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+ 
+-# CMake
+ if("${CMAKE_BUILD_TYPE}" STREQUAL "")
+    set(CMAKE_BUILD_TYPE     RelWithDebInfo)
+ endif("${CMAKE_BUILD_TYPE}" STREQUAL "")
+ set(CMAKE_COLOR_MAKEFILE ON)
+-cmake_minimum_required(VERSION 2.6 FATAL_ERROR)
++
++cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:24]: extra

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 10]
```diff
@@ -0,0 +1,787 @@
...
++if(POLICY CMP0057)
++    cmake_policy(SET CMP0057 NEW)
++endif()
++if(POLICY CMP0011)
++    cmake_policy(SET CMP0011 NEW)
++endif()
++if(POLICY CMP0099)
++    cmake_policy(SET CMP0099 NEW)
++endif()
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:33]: Please explain why you are adjusting the policies, these do not seem required

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 11]
```diff
@@ -0,0 +1,787 @@
...
++set(as_subproject LibFTDI)
++macro(find_package)
++    if(NOT ${ARGV0} IN_LIST as_subproject)
++        _find_package(${ARGV})
++    endif()
++endmacro()
++
+ # Debug build
+-message("-- Build type: ${CMAKE_BUILD_TYPE}")
++message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:48]: extra

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 12]
```diff
@@ -0,0 +1,787 @@
...
++if(POLICY CMP0099)
++    cmake_policy(SET CMP0099 NEW)
++endif()
+ 
+ add_definitions(-Wall)
+ 
+ include(CMakeOptions.txt)
+ 
++set(as_subproject LibFTDI)
++macro(find_package)
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:40]: Overriding the cmake function? 😕 Why is this needed?

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 13]
```diff
@@ -0,0 +1,787 @@
...
+ if ( PYTHON_BINDINGS )
+@@ -150,6 +168,9 @@ if ( EXAMPLES )
+ endif ()
+ add_subdirectory(packages)
+ if ( BUILD_TESTS )
++  project(libftdi1 C CXX)
++  find_package(Boost COMPONENTS unit_test_framework REQUIRED)
++  enable_testing()
+   add_subdirectory(test)
+ endif ()
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:97]: CCI should not be building tests!

extra

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 14]
```diff
@@ -0,0 +1,787 @@
...
++  LIBFTDI_INCLUDE_DIR
++  LIBFTDI_INCLUDE_DIRS
++  LIBFTDI_LIBRARY_DIRS
++NO_CHECK_REQUIRED_COMPONENTS_MACRO
++)
++write_basic_package_version_file (
++LibFTDI1ConfigVersion.cmake
++VERSION ${LIBFTDI_VERSION_STRING}
++COMPATIBILITY AnyNewerVersion
++)
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:156]: This seems like an overkill. The `if` is always true in CCI and should not need to be changed

Is this really required?

**Review by prince-chrismc — CHANGES_REQUESTED** [review]
Try aim to keep the patches to a minimum. These modification make them less portable to newer versions and the recipe harder to maintain...

#3903 and the related PR have a lot of discussion on the topic

I feel this patch is far larger than is required and there's a lot of non trivial changes. I stopped marking but I scanned over the file.

➡️ Please reduce this patch to only the essential changes

**Comment by planetmarshall** [comment]
> Try aim to keep the patches to a minimum. These modification make them less portable to newer versions and the recipe harder to maintain...
> 
> #3903 and the related PR have a lot of discussion on the topic

I took a quick look at the discussion but it seemed to pretty personal and argumentative. In short though I think the patch meets danimtb's criteria in that it only modifies the build system.

> 
> I feel this patch is far larger than is required and there's a lot of non trivial changes. I stopped marking but I scanned over the file.

While a smaller patch would work for CCI, it would not be sufficient to build the tests and examples which would be necessary for the patch to be accepted upstream, which was mentioned above as a goal for using patch files.

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch (hunk 10 above)
Comment (planetmarshall) [reply recipes/libftdi/1.x/patches/0001-cmake-targets.patch:33]: If the policies are not set, there are warnings in the CMake log for recent versions.

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch (hunk 12 above)
Comment (planetmarshall) [reply recipes/libftdi/1.x/patches/0001-cmake-targets.patch:40]: This allows the library to be consumed by examples and applications in the build tree as if it were an external project, and is considered CMake best practice. See [Effective CMake by Daniel Pfeiffer (CPPNow 2017) Slide 31](https://github.com/boostcon/cppnow_presentations_2017/blob/master/05-19-2017_friday/effective_cmake__daniel_pfeifer__cppnow_05-19-2017.pdf)

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch (hunk 13 above)
Comment (planetmarshall) [reply recipes/libftdi/1.x/patches/0001-cmake-targets.patch:97]: It isn't. `BUILD_TESTS` is not defined by the cmake configuration in the conanfile.

**Comment by uilianries** [comment]
> While a smaller patch would work for CCI, it would not be sufficient to build the tests and examples which would be necessary for the patch to be accepted upstream, which was mentioned above as a goal for using patch files

@planetmarshall  Is it an official patch provided by upstream? Or did you create it and pushed to upstream? If it's official, no problem accepting. Otherwise, let's reduce it, testing and examples anre not used on CCI

**Comment by planetmarshall** [comment]
> ...let's reduce it, testing and examples anre not used on CCI

Sure, will take a look over the weekend. I've submitted the patch upstream but could be a while before seeing any feedback.

**Review by danimtb — APPROVED** [review]

**Review by madebr — APPROVED** [review]

**Review by jgsogo — COMMENTED** [review]
I agree with @prince-chrismc that the patch should be simplified. Here the aim should be to modify the build-system the minimum just to make it buildable using Conan... OTH I agree it is better to merge the recipe and improve it in the following iterations.

50/50, wdyt, @prince-chrismc ?

File: recipes/libftdi/1.x/patches/0001-cmake-targets.patch [hunk 15]
```diff
@@ -0,0 +1,787 @@
...
+diff --git a/src/ftdi.c b/src/ftdi.c
+index f5b7018..03d7b07 100644
+--- a/src/ftdi.c
++++ b/src/ftdi.c
+@@ -1328,7 +1328,7 @@ static int ftdi_to_clkbits_AM(int baudrate, unsigned long *encoded_divisor)
+    AM Type chips have only four fractional subdivisors at value[15:14]
+    for subdivisors 0, 0.5, 0.25, 0.125
+ */
+-static int ftdi_to_clkbits(int baudrate, unsigned int clk, int clk_div, unsigned long *encoded_divisor)
++static int ftdi_to_clkbits(int baudrate, int clk, int clk_div, unsigned long *encoded_divisor)
```
Comment (prince-chrismc) [inline recipes/libftdi/1.x/patches/0001-cmake-targets.patch:730]: Actually digging in the patch I found code changes which I was not expect ... this is beyond "fixing cmake"... 700+ lines of diff and lots of meaningless changes

**Review by prince-chrismc — COMMENTED** [review]
There's too many changes in the patch which likely change the output so I do feel comfortable approving... but that's why we have lots of reviewers! 

Given the age there's a few conventions that have changed too if we move this forwarded

**Review by SSE4 — APPROVED** [review]
